core/domain/acl_decorators.py

```diff
@@ -830,6 +831,113 @@ def test_can_edit(self, topic_id, **kwargs):
     return test_can_edit
 
 
+def can_edit_question(handler):
+    """Decorator to check whether the user can edit given question."""
+    def test_can_edit(self, question_id, **kwargs):
+        if not self.user_id:
+            raise base.UserFacingExceptions.NotLoggedInException
+
+        question_rights = question_services.get_question_rights(question_id)
+        if question_rights is None:
+            raise base.UserFacingExceptions.PageNotFoundException
+
+        if question_services.check_can_edit_question(
+                self.user, question_rights):
+            return handler(self, question_id, **kwargs)
+        else:
+            raise self.UnauthorizedUserException(
+                'You do not have credentials to edit this question.')
+    test_can_edit.__wrapped__ = True
+
+    return test_can_edit
+
+
+def can_view_any_question_editor(handler):
+    """Decorator to check whether the user can view any question editor."""
+
+    def test_can_view_any_question_editor(self, **kwargs):
+        if not self.user_id:
+            raise self.NotLoggedInException
+
+        user_actions_info = user_services.UserActionsInfo(self.user_id)
+
+        if (role_services.ACTION_VISIT_ANY_QUESTION_EDITOR in
+                user_actions_info.actions):
+            return handler(self, **kwargs)
+        else:
+            raise self.UnauthorizedUserException(
+                '%s does not have enough rights to access the questions editor'
+                % self.user_id)
+    test_can_view_any_question_editor.__wrapped__ = True
+
+    return test_can_view_any_question_editor
+
+
+def can_delete_question(handler):
+    """Decorator to check whether the user can delete a question."""
+
+    def test_can_delete_question(self, question_id, **kwargs):
+        if not self.user_id:
+            raise self.NotLoggedInException
+
+        user_actions_info = user_services.UserActionsInfo(self.user_id)
+
+        if role_services.ACTION_EDIT_ANY_QUESTION in user_actions_info.actions:
```

Comment:
What's the difference between a person who can edit any question and view any question editor? Also, I don't think anyone can edit any question. We would be using a suggestion based system where an admin can suggest edits but not actually edit them.

Reply:
Umm, I think decided to give admins and TMs editing rights as well.

Reply:
Just a personal opinion here. I think it would be ok to have admins and TMs edit any question. That way, say the question has a minor typo or something like that, the admin can directly make the change and accept the question rather than waiting for the author to get back with the edits. Would help speed up the creation flow a bit.

But then this editing would happen before accepting (i.e, when it is still a suggested question). So this would be more of an can edit suggestion permission rather than a direct can edit question permission. Once published I'm not sure who will have rights to update (edit) a question.

Reply:
I don't have a strong opinion here, but I think you should get feedback from @tjiang11 :)

Reply:
Here's what I had in mind: once a question is published, it no longer belongs to the original creator anymore, in that they can no longer edit it (however, they can see what questions they've gotten approved just for record-keeping convenience). Only admins/TMs can edit a published question. I think the reason for this is, we would expect questions to come from many different people eventually, since to get a large question base, we would need a large amount of people contributing questions. However, a lot of these people contributing questions will just be people who drop in and maybe never come back to Oppia after some time. It will be difficult for us to have the questions remain under the ownership of all these different contributors, since we can't be sure that they'll stick around really, nor do we expect them to. So, with that in mind, I think all published questions should only be editable by admins and TMs. And we don't necessarily want the original authors of questions to be able to edit published questions, because we could have some foul players who publish initially good questions and then spoil them with bad or inappropriate content that we don't endorse, and we can't trust these people since they're just random people we don't necessarily know submitting questions. So at most we would allow them to suggest edits to a published question (but we're planning to allow anyone to suggest edits anyway).

Reply:
Yeah, that sounds fine. So, before publishing, we can allow editing for admins, TMs and the creator, right? As, even if we don't give editing rights to admin/TMs, they could just publish it and edit it anyways.

Reply:
Here's what I suggest:
- Creators will have edit rights before sending the question to review. 
- Once the question is sent for only Admins and TMs will have edit rights. Now admins may reject or approve.
- Once approved/published, editing rights will remain with Admins and TMs.
- If Admins/TMs reject with a comment, editing rights will be transferred to Editors only. 
- Editors may now make changes and send for review again or may delete the question.